File: adventofcode/2023/day09/day09.go
package main

import (
	"bufio"
	"fmt"
	"os"
	"strconv"
	"strings"
	"time"
)

func readInput() [][]int {
	allnums := [][]int{}
	scanner := bufio.NewScanner(os.Stdin)
	for scanner.Scan() {
		line := scanner.Text()
		numss := strings.Split(line, " ")
		nums := []int{}
		for _, ns := range numss {
			n, _ := strconv.Atoi(ns)
			nums = append(nums, n)
		}
		allnums = append(allnums, nums)
	}
	return allnums
}

func diffs(nums []int) []int {
	res := []int{}
	for i := 1; i < len(nums); i++ {
		res = append(res, nums[i]-nums[i-1])
	}
	return res
}

func next(nums []int) int {
	ndiffs := diffs(nums)
	allZeroes := true
	for _, nd := range ndiffs {
		if nd != 0 {
			allZeroes = false
			break
		}
	}
	if allZeroes {
		return nums[0]
	}

	return nums[len(nums)-1] + next(ndiffs)
}

func part1(allnums [][]int) int {
	res := 0
	for _, nums := range allnums {
		res += next(nums)
	}
	return res
}

func part2(allnums [][]int) int {
	res := 0
	for _, nums := range allnums {
		for i, j := 0, len(nums)-1; i < j; i, j = i+1, j-1 {
			nums[i], nums[j] = nums[j], nums[i]
		}
		res += next(nums)
	}
	return res
}

func main() {
	s := time.Now()
	lines := readInput()
	s1 := time.Now()
	fmt.Println(part1(lines))
	e1 := time.Since(s1)
	s2 := time.Now()
	fmt.Println(part2(lines))
	e2 := time.Since(s2)
	e := time.Since(s)
	fmt.Printf("Time for part1: %s, part2: %s, total: %s\n", e1, e2, e)
}

File: adventofcode/2023/day09/day09.zig
const std = @import("std");
const time = std.time;
const print = std.debug.print;
const ArrayList = std.ArrayList(i32);

fn diffs(nums: ArrayList) ArrayList {
    var res = ArrayList.init(std.heap.page_allocator);
    for (1..nums.items.len) |i| {
        res.append(nums.items[i] - nums.items[i - 1]) catch unreachable;
    }
    return res;
}

fn predictNext(nums: ArrayList) i32 {
    const ndiffs = diffs(nums);
    const allZeroes = for (ndiffs.items) |ndiff| {
        if (ndiff != 0) break false;
    } else true;
    if (allZeroes) return nums.items[0];

    return nums.items[nums.items.len - 1] + predictNext(ndiffs);
}

fn predictPrev(nums: ArrayList) i32 {
    const ndiffs = diffs(nums);
    const areZeroes = for (ndiffs.items) |ndiff| {
        if (ndiff != 0) break false;
    } else true;
    if (areZeroes) return nums.items[0];

    return nums.items[0] - predictPrev(ndiffs);
}

pub fn main() !void {
    const s = time.microTimestamp();
    const stdin = std.io.getStdIn().reader();
    var p1: i32 = 0;
    var p2: i32 = 0;
    while (true) {
        var buf: [300]u8 = undefined;
        const line = try stdin.readUntilDelimiterOrEof(&buf, '\n');
        if (line == null) break;

        const l = line.?;
        var nit = std.mem.splitScalar(u8, l, ' ');
        var nums = ArrayList.init(std.heap.page_allocator);
        while (nit.next()) |ns| {
            if (!std.ascii.isASCII(ns[0])) break;

            const n = try std.fmt.parseInt(i32, ns, 10);
            try nums.append(n);
        }
        p1 += predictNext(nums);
        p2 += predictPrev(nums);
    }
    print("{d}\n{d}\n", .{ p1, p2 });
    const e = time.microTimestamp();
    print("Time taken: {s}\n", .{std.fmt.fmtDurationSigned(e - s)});
}
